ENH: Add encoding option to numpy text IO

juliantaylor opened the pull request:
add encoding flag to np.loadtxt to be able to load non default encoded
text files.

charris commented:
@juliantaylor Could you revisit this when you finish with higher priority stuff. Might be easier with support for 2.5 dropped. Also interested in @pv comment about text files.

charris commented:
Pushing this off (again) to 1.11.

charris commented:
@juliantaylor Needs a rebase.

charris commented:
@juliantaylor Still interested in this?

ChrisBarker-NOAA commented:
I hope so -- this would be nice :-)

charris commented:
@juliantaylor Closing this. Please resubmit if you get the urge to continue. Anyone else interested in this is welcome to pull the code and give it a shot.

juliantaylor committed:
attempt to salvage loadtxt and genfromtxt

juliantaylor committed:
update some internal tests

juliantaylor commented:
rebased on current head, it is probably impossible to salvage these functions without breaking backward compatibility.
One of the bigger issues are the converters, they can try to decode the bytes, switching the text io to actually do text io breaks this so we need to convert to text back to bytes again before passing them to the user converters. Possibly a new argument could be used to disable this for new code.

pv commented:
You can perhaps use encoding='bytes' for the bytestring behavior assumed 
in the original port.

juliantaylor committed:
update datasource

use TextIOWrapper for compressed files on python2

juliantaylor committed:
less ugly dtype=S hack

juliantaylor committed:
use special value encoding="bytes" for converter value

juliantaylor committed:
windows test fix

eric-wieser commented on numpy/lib/_datasource.py:
```diff
@@ -38,9 +38,44 @@
+    if "t" in mode:
+        raise UnsupportedMode("bz2 files not supported in python2")
+    else:
+        return bz2.BZ2File(fn, mode.replace("t", ""))
```
This `replace` is always a no op, isn't it?

juliantaylor replied:
yes
bz2 probably can't easily work in python2 as it is not compatible with TextIOWrapper. Luckily it is not very much used anymore.

juliantaylor committed:
add gzip line separator test

juliantaylor commented:
Good idea, I used encoding='bytes' to activate the converter workaround.
This might be doable after all, I have updated the datasource to support encoding and newline also for gzip via TextIOWrapper. This might also fix some line separator issues.
Also found a nicer workaround for requested byte dtype ('S') in loadtxt.

still todo:

- [ ] cleanup, docstrings

- [ ] getpreferredencoding workaround for binary input stream in genfromtxt

- [ ] more tests (decoding converter in loadtxt, bz2 tests in python3, encoding tests for genfromtxt)

- [ ] savetxt

- [ ] xz support for py3

juliantaylor committed:
try to keep genfromtxt backward compat

encode strings back to bytes on dtype=None input
only use unicode if this cannot be done

juliantaylor committed:
add decoding converter tests

juliantaylor committed:
support unicode user dtype in genfromtxt

juliantaylor commented:
I guess most stuff is in, genfromregex is mission but that probably is not used as much.
supporting binary input streams that result in decoded unicode in the resulting array via getpreferredencoding is probably not reasonable to implement for genfromtxt, but it is more or less buggy usage, the only know usage of it is in loadtxt which works (somewhat).

I'll put out a call for testing tomorrow. General review is appreciated, but cleanup and stiyle will follow after testing.
Given how long overdue this fix is I think this should be in 1.13 so we can put deprecation warnings on the workarounds sooner.

juliantaylor committed:
add unicode support to fromregexp

juliantaylor committed:
add docstrings

[ci skip]

eric-wieser commented on numpy/lib/_iotools.py:
```diff
@@ -236,6 +242,12 @@ def _fixedwidth_splitter(self, line):
+            if self.encoding is None:
+                line = line.decode('latin1')
+            else:
+                line = line.decode(self.encoding)
```
These 5 lines probably deserve a helper method

eric-wieser commented on numpy/lib/_iotools.py:
```diff
@@ -529,7 +541,8 @@ class StringConverter(object):
                     (nx.longdouble, nx.longdouble, nx.nan),
-                    (nx.string_, bytes, b'???')])
+                    (nx.unicode_, asunicode, '???'),
+                    (nx.string_, asbytes, '???')])
```
These are a little sketchy, as they both force `latin1` as the encoding

juliantaylor replied:
I don't like it either, but I have not succeeded in creating a testcase that fails here yet.

juliantaylor committed:
move line decoding to a function

juliantaylor committed:
add binary stream decode test

juliantaylor committed:
only use writewrap in savetxt when necessary

juliantaylor committed:
don't convert data to lists unnecessarily in genfromtxt

juliantaylor committed:
avoid the file encoding workaround if encoding is provided by user

juliantaylor committed:
add xz support and add tests

juliantaylor commented:
I think this should be ready.

mhvk commented on numpy/lib/_datasource.py:
```diff
@@ -38,9 +38,44 @@
+    if "t" in mode:
+        raise ValueError("bz2 files not supported in python2")
+    else:
+        return bz2.BZ2File(fn, mode.replace("t", ""))
```
Since you just checked that `'t' not in mode`, I don't understand the understand the need for the replace.

juliantaylor replied:
a remnant of copy pasting the code from python3, can be removed.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -885,23 +900,26 @@ def loadtxt(fname, dtype=float, comments='#', delimiter=None,
     X = []
 
+    # input may be a python2 io stream
+    if encoding is not None and encoding != 'bytes':
```
The test for `bytes` is superfluous, since for that case `encoding` has already be set to `None`

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1006,7 +1020,18 @@ def split_line(line):
+            if byte_converters:
+                # converters may use decode to workaround numpy's oldd behaviour,
+                # so encode the string again before passing to the user converter
+                def tobytes_first(x, conv):
```
This seems to redefine `asbytes` from `py3k.py`; can one just use that? I.e., something like
```
converters[i] = lambda x: conv(asbytes(x))
```

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1006,7 +1020,18 @@ def split_line(line):
+                user_conv = functools.partial(tobytes_first, conv=conv)
+                converters[i] = user_conv
+            else:
+                converters[i] = conv
```
Maybe here put the check on `conv is bytes` from below? It seems nicer to just ensure the converters are all set up correctly at this stage.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1020,8 +1045,11 @@ def split_line(line):
                 raise ValueError("Wrong number of columns at line %d"
                                  % line_num)
 
+            converters = [conv if conv != bytes else
```
See comment above: maybe best to do the check while constructing `converters`

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1179,21 +1212,51 @@ def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
+        """ convert to unicode in py2 or to bytes on bytestream inputs """
+        def __init__(self, fh, encoding):
+            self.fh = fh
+            self.encoding = encoding if encoding else 'latin1'
```
I don't think this class should set that default; maybe rather do it in the call to it?

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1179,21 +1212,51 @@ def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
-                fh = open(fname, 'w')
+        # need to convert str to unicode for text io output
+        if sys.version_info[0] == 2:
+            fh = WriteWrap(fh, encoding)
```
So I'd put `encoding or 'latin1'` here (and below)

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1321,16 +1388,22 @@ def fromregex(file, regexp, dtype):
+        if isinstance(content, bytes) and not isinstance(regexp, bytes):
+            regexp = asbytes(regexp)
+        elif not isinstance(content, bytes) and isinstance(regexp, bytes):
+            regexp = regexp.decode('latin1')
```
Can the `regexp` be anything but `str`? If that is the case, I'd still use `regexp = asstr(regexp)` here to keep things localized to py3k.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1539,16 +1611,18 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
             "The input argument 'converter' should be a valid dictionary "
             "(got '%s' instead)" % type(user_converters))
 
+    byte_converters = False
```
In `else` clause, as for `loadtxt`.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1642,9 +1716,11 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
     # Rename missing_values for convenience
     user_missing_values = missing_values or ()
+    if isinstance(user_missing_values, bytes):
+        user_missing_values = user_missing_values.decode('latin1')
```
Can this happen?  If so, use `asstr` instead of explicit decoding

charris replied:
`asstr` does seem better.

charris replied:
Chnged my mind.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1772,11 +1848,23 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
         else:
             testing_value = None
-        converters[i].update(conv, locked=True,
+        user_conv = conv
```
I'd put this in an `else` clause for clarity.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1772,11 +1848,23 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
+                    return conv(x)
+                return conv(x.encode("latin1"))
+            import functools
+            user_conv = functools.partial(tobytes_first, conv=conv)
```
Same as for `loadtxt`: `user_conv = lambda x: conv(asbytes(x))` (if that indeed works)

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1893,16 +1981,39 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
+                for j in range(len(data)):
+                    row = list(data[j])
+                    for i in strcolidx:
+                        row[i] = row[i].encode('latin1')
```
use `asbytes`

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1893,16 +1981,39 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
+                    row = list(data[k])
+                    for i in strcolidx:
+                        if isinstance(row[i], bytes):
+                            row[i] = row[i].decode('latin1')
```
Use `asstr`

mhvk commented on numpy/lib/tests/test__iotools.py:
```diff
@@ -134,10 +134,9 @@ def test_validate_wo_names(self):
-        return date(*time.strptime(s.decode('latin1'), "%Y-%m-%d")[:3])
-    else:
-        return date(*time.strptime(s, "%Y-%m-%d")[:3])
+    if type(s) == bytes:
```
Again, use `asstr`

mhvk commented on numpy/lib/tests/test_io.py:
```diff
@@ -52,10 +73,9 @@ def strptime(s, fmt=None):
-        return datetime(*time.strptime(s.decode('latin1'), fmt)[:3])
-    else:
-        return datetime(*time.strptime(s, fmt)[:3])
+    if type(s) == bytes:
```
Use `asstr`

mhvk reviewed (commented):
Overall, this looks great! My main comment is to use `asstr` and `asbytes` as much as possible, so that the choice of `latin1` as an encoding remains localized to `py3k`.

Indeed, there may be an argument for endowing those functions with an `encoding` argument, and for making the "default" overall encoding settable, but that is probably better done separately.

juliantaylor committed:
cleanup compressed file handling in datasource

juliantaylor committed:
remove two now unnecessary abstractions

juliantaylor committed:
fix encoding argument not being passed to Linesplitter

juliantaylor committed:
move decoding into Linesplitter's handyman function

juliantaylor committed:
cleanup

juliantaylor commented:
updated.
asbytes and asstr mostly cannot be used as the new code is based on decoding everything to unicode and encoding back to back to bytes in the legacy mode. These functions do not do that in python2.
Using latin1 should be fine for the back decoding in python2 as it overlaps with ascii and thus only increases the range of characters the functions can handle.

But this all does imply increased memory usage in python2 due to 2-4 byte unicode strings everywhere.

eric-wieser commented on numpy/lib/_datasource.py:
```diff
@@ -38,13 +38,61 @@
+    import bz2
+
+    if "t" in mode:
+        raise ValueError("bz2 text files not supported in python2")
```
This should probably have a:

```
if "b" in mode:
    raise ValueError("Invalid mode: %r" % (mode,))
```

In fact, can this validation be extracted to a helper function, since it appears twice?

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -1953,6 +2066,8 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
                         ishomogeneous &= (ttype == dtype.type)
                         if ttype == np.string_:
```
Nit: change this line to `np.bytes_` for the sake of being explicit?

juliantaylor committed:
ENH: change loadtxt to use a generator to load data

Load data in chunks and fill it into an array grown with resize.
This significantly reduces the memory consumption of the function.

mhvk commented on numpy/lib/npyio.py:
```diff
@@ -1772,11 +1849,24 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
         else:
             testing_value = None
-        converters[i].update(conv, locked=True,
+        if conv == bytes:
```
True nitpick: class comparison using `is` instead of `==`.

juliantaylor commented:
we could recover some of the memory increase by switching to using generators which should be simple for loadtxt. For genfromtxt it might be a bit more tricky.

mhvk commented:
I'm not sure if it is worth it, nor really if it is possible, but would an alternative to going from/to unicode in python2 for `encoding='bytes'` be to not make any conversion at all, i.e., explicitly rely on the equivalency between str and bytes in python2? (This basically presumes that all the `b` were inserted in front of strings only to make things "work" on python3, so that their removal in the current PR doesn't really affect python2 directly.)

juliantaylor commented:
The goal of this is to add the encoding flag and handling input text io streams instead of binary io streams only. Doing so in python 2 means the code needs to deal with unicode strings and bytes.
It is easier to just convert everything to unicode instead of adding two code paths for everything.

Treating bytes and strings the same in python2 is what got us into this mess in the first place. To avoid having to add two code path, it used to take all strings and convert them to bytes in python 3, this PR turns this around and turns all bytes into strings in python2 instead which is imo much more sane.

mhvk commented:
@juliantaylor - totally agree that converting everything to unicode is much more sane; it was only vague wondering whether having an "i don't care, just pass things along" interpretation of `encoding=''bytes"` would work for python2. I guess the main objective would be to have code that is as clear as possible on python3, and works on python2.

Anyway, the current PR looks good to me!

juliantaylor commented:
I guess if we do not add unicode as an option to the `StringConverter` in bytes encoding mode it might work. But I'm not sure if it is really worth the extra code.

juliantaylor commented:
Changing StringConverter seems tricky as it is based on global mappings. But for loadtxt we could have non-decoding version for bytes encoding quite easily. Changing that to be generator based seems more elegant to me, I don't think the extra decoding step should be too costly in terms of performance.

juliantaylor commented:
do we still want to put this in 1.13? If so it probably just needs deprecation warnings for byte stream input without encoding and maybe a warning for string -> bytes backward compatiblity conversion

mhvk commented:
Would be wonderful to have in 1.13!

homu commented:
:umbrella: The latest upstream changes (presumably #9075) made this pull request unmergeable. Please resolve the merge conflicts.

juliantaylor committed:
DOC: add release notes for text IO changes

juliantaylor committed:
DEPR: add a deprecation warning when reading strings without encoding

juliantaylor committed:
add test for savetxt into StringIO

juliantaylor commented:
added release notes and a deprecation warning for reading strings without encoding set and using autodtype.
This should go in soon, should I squash the commits or keep the history with some minor cleanup?

eric-wieser commented on numpy/lib/_datasource.py:
```diff
@@ -65,27 +111,39 @@ class _FileOpeners(object):
         try:
             import bz2
-            self._file_openers[".bz2"] = bz2.BZ2File
+            if sys.version_info[0] >= 3:
```
Can we use `version_info.major` here instead?

eric-wieser replied:
Nevermind, we do this everywhere else, so from a consistency perspective, lets leave this

eric-wieser commented on numpy/lib/_iotools.py:
```diff
@@ -189,12 +189,10 @@ def autostrip(self, method):
     #
 
-    def __init__(self, delimiter=None, comments=b'#', autostrip=True):
+    def __init__(self, delimiter=None, comments='#', autostrip=True, encoding=None):
```
Should this be `u'#'`?

charris replied:
Hmm, looks like it if we work in unicode. The `u` prefix came in late (python 3.4 ?), but is now available in all the versions we support.

charris replied:
I think we are OK, in Python 2.7
```
In [1]: u'123' + '4'
Out[1]: u'1234'

In [2]: '3' in u'123'
Out[2]: True
```
So the strings are promoted to unicode. Narrow builds (ucf-2 unicode), might be a problem.

eric-wieser commented on numpy/lib/tests/test__iotools.py:
```diff
@@ -20,61 +20,61 @@ class TestLineSplitter(TestCase):
+        strg = "  1     3  4  5  6# test"
         test = LineSplitter((6, 6, 9))(strg)
-        assert_equal(test, [b'1', b'3  4', b'5  6'])
+        assert_equal(test, ['1', '3  4', '5  6'])
```
Similarly, should all these tests be using `unicode` objects in py2? Can we add some tests on the returned string type, deciding whether this should return `unicode` or `bytes`? AFAIK, on py2 `assert_equal` treats str and unicode as equal.

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -885,22 +904,24 @@ def loadtxt(fname, dtype=float, comments='#', delimiter=None,
+    # TOOD emit portability warning?
+    elif fencoding is None:
+        import locale
+        fencoding = locale.getpreferredencoding()
```
This can only happen if we passed `encoding=None` to `np.lib._datasource.open` or `fname.encoding is None`, right?

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -1951,8 +2097,10 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
+                        if ttype is np.bytes_:
                             ttype = "|S%i" % max(len(row[i]) for row in data)
+                        elif ttype is np.unicode_:
+                            ttype = "|U%i" % max(len(row[i]) for row in data)
```
I think these lines would be better as

````
if issubdtype(ttype, np.character):
    ttype = (ttype, max(len(row[i]) for row in data))
```

charris commented:
Time to get this finished up. @juliantaylor Are you available? If not I will fork this into another PR.

juliantaylor commented:
I can probably look this over again in the next few days

charris commented:
@juliantaylor  Ready for review?

charris commented on doc/release/1.14.0-notes.rst:
```diff
@@ -9,6 +9,8 @@ Highlights
 
 * The `np.einsum` function will use BLAS when possible
+* ``genfromtxt``, ``loadtxt``, ``fromregex`` and ``savetxt`` can now handle files
+with arbitrary encoding supported by Python.
```
Needs indentation

charris commented on numpy/lib/_datasource.py:
```diff
@@ -38,13 +38,59 @@
 _open = open
 
+def _check_mode(mode, encoding, newline):
+    if "t" in mode:
```
Needs docstring.

charris commented on numpy/lib/_datasource.py:
```diff
@@ -38,13 +38,59 @@
+            raise ValueError("Argument 'newline' not supported in binary mode")
+
+def _python2_bz2open(fn, mode, encoding, newline):
+    """ wrapper to open bz2 in text mode """
```
Needs docstring of the standard type with expanded explanation and documentation of parameters.

charris commented on numpy/lib/_datasource.py:
```diff
@@ -38,13 +38,59 @@
+        return bz2.BZ2File(fn, mode)
+
+def _python2_gzipopen(fn, mode, encoding, newline):
+    """ wrapper to open gzip in text mode """
```
Needs docstring.

charris commented on numpy/lib/_datasource.py:
```diff
@@ -115,7 +173,7 @@ def __getitem__(self, key):
 
 _file_openers = _FileOpeners()
```
This singleton seems a bit odd. Old design I expect.

charris commented on numpy/lib/_iotools.py:
```diff
@@ -189,12 +189,10 @@ def autostrip(self, method):
-        if isinstance(delimiter, unicode):
-            delimiter = delimiter.encode('ascii')
-        if (delimiter is None) or _is_bytes_like(delimiter):
+        if (delimiter is None) or isinstance(delimiter, basestring):
```
So in Python 2 either ascii or unicode?

charris replied:
Probably OK.

charris commented on numpy/lib/npyio.py:
```diff
@@ -1190,21 +1250,51 @@ def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
         fmt = asstr(fmt)
     delimiter = asstr(delimiter)
 
+    class WriteWrap(object):
```
Need better docstring.

charris commented on numpy/lib/npyio.py:
```diff
@@ -960,21 +981,43 @@ def pack_items(items, packing):
             return []
 
+    def read_data(chunk_size):
+        # Parse each line, including the first
```
Needs better docstring, in paricular `chunc_size`.

charris commented on numpy/lib/npyio.py:
```diff
@@ -1460,6 +1562,15 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
 
         .. versionadded:: 1.10.0
+    encoding: string, optional
+        Encoding used to decode the inputfile. Does not apply to input streams.
```
What is a stream in this context?

charris commented:
New functions need docstrings. Also, some of the comments look useful.

charris commented:
Just to be clear, are the assumptions here that:

1) Input files are text, default encoding is ascii (or latin1)?
2) String data is str in Python 2 and python 3 for default encoding (S).
3) String data is unicode (U) for other encoding?
4) Output files are text (unicode).

charris commented:
Rebased and squashed in #10054, so closing this.
